# Segmenting and Clustering Neighborhoods in Toronto

In this notebook, we will explore, segment, and cluster the neighborhoods in the city of Toronto. We will end up with groups of similar neighborhoods based on the categories of the venues in each neighborhood. 

For the Toronto neighborhood data, a Wikipedia page exists that has information we need to explore and cluster the neighborhoods in Toronto. It contains the name of each neighborhood along with the borough and the postal code of the neighborhood.

## Part 1: Scraping the Data

### Importing libraries

In [1]:
import pandas as pd
import numpy as np
import folium
import requests
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

In [2]:
pd.set_option('max_rows', 140)

### Reading the table from the Wikipedia page

In [3]:
dfs = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
df = dfs[0]

In [4]:
df.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [5]:
df.shape

(288, 3)

### Making small modifications on the column names

In [6]:
df.columns = ['PostalCode', 'Borough', 'Neighborhood']

### Removing records where the borough is "not assigned"

In [7]:
df = df[df.Borough != 'Not assigned']

In [8]:
df.shape

(211, 3)

### Merging records where multiple neighborhoods share the same borough

For example, there are two neighborhoods (Harbourfront and Regent Park) that share the same borough (Downtown Toronto) as shown below:

In [9]:
df[df.PostalCode == 'M5A']

PostalCode           Borough  Neighborhood
4        M5A  Downtown Toronto  Harbourfront
5        M5A  Downtown Toronto   Regent Park

We will merge records where multiple neighborhoods share the same borough:

In [10]:
df = (df.groupby(['PostalCode', 'Borough'])['Neighborhood']
      .apply(lambda x: "{}".format(', '.join(x))).reset_index())

Now let's take a look at the record of Downtown Toronto borough:

In [11]:
df[df.PostalCode == 'M5A']

PostalCode           Borough               Neighborhood
53        M5A  Downtown Toronto  Harbourfront, Regent Park

We can see that the previous two records were merged into one and the neighborhood names got concatenated.

### Dealing with the case where the neighborhood is "not assigned"

For some records in the table, there is a valid value for the borough but the neighborhood has the value "Not assigned" as shown below:

In [12]:
df[df.Neighborhood.str.contains('Not assigned')]

PostalCode       Borough  Neighborhood
85        M7A  Queen's Park  Not assigned

In this case, we will give the borough name to the neighborhood:

In [13]:
df.loc[df.Neighborhood.str.contains('Not assigned'), 'Neighborhood'] = \
    df.loc[df.Neighborhood.str.contains('Not assigned'), 'Borough']

Now let's take a look at the previous record:

In [14]:
df[df.Borough == "Queen's Park"]

PostalCode       Borough  Neighborhood
85        M7A  Queen's Park  Queen's Park

We can see that the neighborhood got the same name as the borough.

### The shape of the final neighborhood table

In [15]:
df.shape

(103, 3)

## Part 2: Getting Latitude and Longitude Coordinates of the Neighborhoods

In [16]:
df.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Rouge, Malvern
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

### Reading a file that contains the coordinates of each borough

In [17]:
lat_lng_df = pd.read_csv('https://cocl.us/Geospatial_data')

In [18]:
lat_lng_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### Adding langitude and latitude to our dataframe

In [19]:
df.reset_index(drop=True, inplace=True)
df['Latitude'] = -99999.9
df['Longitude'] = -99999.9

for i in range(df.shape[0]):
    postalcode = df.loc[i, 'PostalCode']
    lat = lat_lng_df.loc[lat_lng_df['Postal Code'] == postalcode, 'Latitude'].squeeze()
    lng = lat_lng_df.loc[lat_lng_df['Postal Code'] == postalcode, 'Longitude'].squeeze()
    df.loc[i, 'Latitude'] = lat
    df.loc[i, 'Longitude'] = lng    

In [20]:
df.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

## Part 3: Clustering of the Neighborhoods

### Creating a map of Toronto with neighborhoods superimposed on top

In [21]:
# Toronto coordinates
latitude = 43.6532
longitude = -79.3832

map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add neighborhood markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], 
                                           df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='#333333',
        fill=True,
        fill_color='#ff5349',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Getting neighborhood data from Foursquare using its API

In [22]:
# Foursquare credentials
CLIENT_ID = 'A1VPXW0G4VZLCKIMZUOGCWBXBGIY5UP21CRGT0R4OJXJOWBQ' 
CLIENT_SECRET = 'FGC4DCS0JRFEC25IFE1NC0AFZM2KDTVRRFUK13VFVNG4DHXC' 
VERSION = '20180605'

In [24]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    """
    A function that retrieves information about venues in each neighborhood.
    It takes as input the names of the neighborhoods, their latitudes and longitudes.
    It returns a dataframe with information about each neighborhood and its venues.
    """
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = ('https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}'
               '&v={}&ll={},{}&radius={}&limit={}'
               .format(CLIENT_ID, CLIENT_SECRET, VERSION, lat, lng, radius, LIMIT))
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(name, lat, lng, v['venue']['name'], 
                             v['venue']['location']['lat'], 
                             v['venue']['location']['lng'],  
                             v['venue']['categories'][0]['name']) 
                            for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 'Neighborhood Latitude', 'Neighborhood Longitude', 
                             'Venue', 'Venue Latitude', 'Venue Longitude', 'Venue Category']
    
    return(nearby_venues)


tor_venues = getNearbyVenues(names=df['Neighborhood'],
                             latitudes=df['Latitude'],
                             longitudes=df['Longitude'])

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, 

In [25]:
tor_venues.shape

(2271, 7)

In [26]:
tor_venues.head()

Neighborhood  Neighborhood Latitude  \
0                          Rouge, Malvern              43.806686   
1  Highland Creek, Rouge Hill, Port Union              43.784535   
2       Guildwood, Morningside, West Hill              43.763573   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.160497            Royal Canadian Legion       43.782533   
2              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
3              -79.188711                G & G Electronics       43.765309   
4              -79.188711                 Big Bite Burrito       43.766299   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.189914           Pizza Place  
3       -79.191537     Electronics Store  
4       -79.190720    Mexican Restaurant

### How many venues were returned for each neighborhood?

In [27]:
tor_venues.groupby('Neighborhood').size().to_frame('Venue count')

Venue count
Neighborhood                                                   
Adelaide, King, Richmond                                    100
Agincourt                                                     4
Agincourt North, L'Amoreaux East, Milliken, Ste...            3
Albion Gardens, Beaumond Heights, Humbergate, J...           10
Alderwood, Long Branch                                        9
Bathurst Manor, Downsview North, Wilson Heights              18
Bayview Village                                               4
Bedford Park, Lawrence Manor East                            25
Berczy Park                                                  55
Birch Cliff, Cliffside West                                   4
Bloordale Gardens, Eringate, Markland Wood, Old...            8
Brockton, Exhibition Place, Parkdale Village                 21
Business Reply Mail Processing Centre 969 Eastern            16
CFB Toronto, Downsview East                                   4
CN Tower, Bathurst Quay, Island airport, Harbou...           16
Cabbagetown, St. James Town                                  49
Caledonia-Fairbanks                                           5
Canada Post Gateway Processing Centre                        11
Cedarbrae                                                     7
Central Bay Street                                           85
Chinatown, Grange Park, Kensington Market                   100
Christie                                                     15
Church and Wellesley                                         92
Clairlea, Golden Mile, Oakridge                               9
Clarks Corners, Sullivan, Tam O'Shanter                      13
Cliffcrest, Cliffside, Scarborough Village West               3
Cloverdale, Islington, Martin Grove, Princess G...            1
Commerce Court, Victoria Hotel                              100
Davisville                                                   33
Davisville North                                              7
Deer Park, Forest Hill SE, Rathnelly, South Hil...           15
Del Ray, Keelesdale, Mount Dennis, Silverthorn                4
Design Exchange, Toronto Dominion Centre                    100
Don Mills North                                               6
Dorset Park, Scarborough Town Centre, Wexford H...            7
Dovercourt Village, Dufferin                                 16
Downsview Central                                             3
Downsview Northwest                                           5
Downsview West                                                5
Downsview, North Park, Upwood Park                            4
East Birchmount Park, Ionview, Kennedy Park                   6
East Toronto                                                  4
Emery, Humberlea                                              2
Fairview, Henry Farm, Oriole                                 67
First Canadian Place, Underground city                      100
Flemingdon Park, Don Mills South                             22
Forest Hill North, Forest Hill West                           4
Glencairn                                                     7
Guildwood, Morningside, West Hill                             7
Harbord, University of Toronto                               33
Harbourfront East, Toronto Islands, Union Station           100
Harbourfront, Regent Park                                    50
High Park, The Junction South                                23
Highland Creek, Rouge Hill, Port Union                        1
Hillcrest Village                                             5
Humber Bay Shores, Mimico South, New Toronto                 14
Humber Bay, King's Mill Park, Kingsway Park Sou...            3
Humber Summit                                                 2
Humewood-Cedarvale                                            4
Kingsview Village, Martin Grove Gardens, Richvi...            4
Kingsway Park South West, Mimico NW, The Queens...           14
L'Amoreaux West     

We notice that there are 99 neighborhoods in the table while we had 103 neighborhoods before we got the data from Foursquare. This means that Foursquare didn't return any venue for 4 neighborhoods. These 4 neighborhoods are:

In [28]:
excluded_neighborhoods = set(df['Neighborhood']).difference(tor_venues['Neighborhood'])
excluded_neighborhoods

{'Islington Avenue', 'Newtonbrook, Willowdale', 'Upper Rouge'}

### Performing one-hot on the venue category variable

In [29]:
# one hot encoding
toronto_onehot = pd.get_dummies(tor_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
# we used "Neighborhood_" instead of just "Neighborhood" because
# there is a venue category  called "Neighborhood"
toronto_onehot['Neighborhood_'] = tor_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood_  Accessories Store  \
0                          Rouge, Malvern                  0   
1  Highland Creek, Rouge Hill, Port Union                  0   
2       Guildwood, Morningside, West Hill                  0   
3       Guildwood, Morningside, West Hill                  0   
4       Guildwood, Morningside, West Hill                  0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   ...  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0  ...              0                              0                 0   
1  ...              0                              0                 0   
2  ...              0                              0                 0   
3  ...              0                              0                 0   
4  ...              0                              0                 0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0            0                      0                0         0            0   
1            0                      0                0         0            0   
2            0                      0                0         0            0   
3            0                      0                0         0            0   
4            0                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 274 columns]

In [30]:
toronto_onehot.shape

(2271, 274)

### Grouping rows by neighborhood and taking the mean of the frequency of occurrence of each category for each neighborhood

In [31]:
toronto_grouped = toronto_onehot.groupby('Neighborhood_').mean().reset_index()
toronto_grouped.head()

Neighborhood_  Accessories Store  \
0                           Adelaide, King, Richmond                0.0   
1                                          Agincourt                0.0   
2  Agincourt North, L'Amoreaux East, Milliken, St...                0.0   
3  Albion Gardens, Beaumond Heights, Humbergate, ...                0.0   
4                             Alderwood, Long Branch                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0                 0.03   
1             0.0              0.0               0.0                 0.00   
2             0.0              0.0               0.0                 0.00   
3             0.0              0.0               0.0                 0.00   
4             0.0              0.0               0.0                 0.00   

   ...  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0  ...            0.0                           0.02               0.0   
1  ...            0.0                           0.00               0.0   
2  ...            0.0                           0.00               0.0   
3  ...            0.0                           0.00               0.0   
4  ...            0.0                           0.00               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0          0.0                    0.0              0.0      0.01          0.0   
1          0.0                    0.0              0.0      0.00          0.0   
2          0.0                    0.0              0.0      0.00          0.0   
3          0.0                    0.0              0.0      0.00          0.0   
4          0.0                    0.0              0.0      0.00          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 274 columns]

In [32]:
toronto_grouped.shape

(100, 274)

### The most common categories for each neighborhood

In [33]:
def return_most_common_venues(row, num_top_cat):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_cat]


num_top_cat = 5
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood_']
for ind in np.arange(num_top_cat):
    try:
        columns.append('{}{} Most Common Category'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Category'.format(ind+1))

# create a new dataframe
neighborhoods_categories_sorted = pd.DataFrame(columns=columns)
neighborhoods_categories_sorted['Neighborhood_'] = toronto_grouped['Neighborhood_']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_categories_sorted.iloc[ind, 1:] = return_most_common_venues(
        toronto_grouped.iloc[ind, :], num_top_cat)

neighborhoods_categories_sorted.head()

Neighborhood_ 1st Most Common Category  \
0                           Adelaide, King, Richmond              Coffee Shop   
1                                          Agincourt                   Lounge   
2  Agincourt North, L'Amoreaux East, Milliken, St...                     Park   
3  Albion Gardens, Beaumond Heights, Humbergate, ...            Grocery Store   
4                             Alderwood, Long Branch              Pizza Place   

  2nd Most Common Category 3rd Most Common Category 4th Most Common Category  \
0                     Café                      Bar               Steakhouse   
1           Breakfast Spot             Skating Rink           Clothing Store   
2               Playground         Sculpture Garden              Yoga Studio   
3     Fast Food Restaurant             Liquor Store      Fried Chicken Joint   
4                      Gym             Dance Studio                 Pharmacy   

  5th Most Common Category  
0           Cosmetics Shop  
1              Yoga Studio  
2                Drugstore  
3           Sandwich Place  
4              Coffee Shop

### Clustering

Now we apply K-means clustering on the dataframe stored in `toronto_grouped` variable which includes the relative frequency of each venue-category for each neighborhood.

In [34]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood_', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 2, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [36]:
# add clustering labels
neighborhoods_categories_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df.rename(columns={'Neighborhood': 'Neighborhood_'}).copy()
toronto_merged = toronto_merged[~toronto_merged['Neighborhood_'].isin(excluded_neighborhoods)]

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_categories_sorted.set_index('Neighborhood_'), on='Neighborhood_')

toronto_merged.head() # check the last columns!

PostalCode      Borough                           Neighborhood_   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Category  \
0 -79.194353               1     Fast Food Restaurant   
1 -79.160497               0                      Bar   
2 -79.188711               0             Intersection   
3 -79.216917               0              Coffee Shop   
4 -79.239476               0                   Bakery   

  2nd Most Common Category 3rd Most Common Category 4th Most Common Category  \
0      Dumpling Restaurant                    Diner           Discount Store   
1              Yoga Studio      Dumpling Restaurant           Discount Store   
2           Breakfast Spot              Pizza Place      Rental Car Location   
3        Korean Restaurant              Yoga Studio           Discount Store   
4         Hakka Restaurant                     Bank     Caribbean Restaurant   

  5th Most Common Category  
0                  Dog Run  
1                  Dog Run  
2        Electronics Store  
3                  Dog Run  
4       Athletics & Sports

### Creating a map that shows the neighborhoods and their clusters
We will create a map that shows a marker for each neighborhood; the color of the marker represents the cluster of the neighborhood.

In [37]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], 
                                  toronto_merged['Neighborhood_'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examining clusters
Let's see the neighborhoods in each of the five clusters:

#### Cluster 1

In [38]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Cluster Labels       1st Most Common Category  \
1         Scarborough               0                            Bar   
2         Scarborough               0                   Intersection   
3         Scarborough               0                    Coffee Shop   
4         Scarborough               0                         Bakery   
5         Scarborough               0                     Playground   
6         Scarborough               0                 Discount Store   
7         Scarborough               0                         Bakery   
8         Scarborough               0            American Restaurant   
9         Scarborough               0                College Stadium   
10        Scarborough               0              Indian Restaurant   
11        Scarborough               0      Middle Eastern Restaurant   
12        Scarborough               0                         Lounge   
13        Scarborough               0                       Pharmacy   
15        Scarborough               0             Chinese Restaurant   
17         North York               0           Fast Food Restaurant   
18         North York               0                 Clothing Store   
19         North York               0                           Café   
22         North York               0                    Coffee Shop   
24         North York               0                    Coffee Shop   
26         North York               0            Japanese Restaurant   
27         North York               0                            Gym   
28         North York               0                    Coffee Shop   
29         North York               0         Furniture / Home Store   
30         North York               0                           Park   
31         North York               0                  Grocery Store   
32         North York               0                 Baseball Field   
33         North York               0                 Discount Store   
34         North York               0          Portuguese Restaurant   
35          East York               0           Fast Food Restaurant   
36          East York               0                       Bus Stop   
37       East Toronto               0                          Trail   
38          East York               0                    Coffee Shop   
39          East York               0              Indian Restaurant   
40          East York               0                    Pizza Place   
41       East Toronto               0               Greek Restaurant   
42       East Toronto               0                           Park   
43       East Toronto               0                           Café   
45    Central Toronto               0                           Park   
46    Central Toronto               0                 Clothing Store   
47    Central Toronto               0                   Dessert Shop   
48    Central Toronto               0                     Restaurant   
49    Central Toronto               0                    Coffee Shop   
51   Downtown Toronto               0                    Coffee Shop   
52   Downtown Toronto               0                    Coffee Shop   
53   Downtown Toronto               0                    Coffee Shop   
54   Downtown Toronto               0                    Coffee Shop   
55   Downtown Toronto               0                    Coffee Shop   
56   Downtown Toronto               0                    Coffee Shop   
57   Downtown Toronto               0                    Coffee Shop   
58   Downtown Toronto               0                    Coffee Shop   
59   Downtown Toronto               0                    Coffee Shop   
60   Downtown Toronto               0                    Coffee Shop   
61   Downtown Toronto               0                    Coffee Shop   
62         North York               0             Italian Restaurant   
64    Central Toronto               0                 

#### Cluster 2

In [39]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Category  \
0  Scarborough               1     Fast Food Restaurant   

  2nd Most Common Category 3rd Most Common Category 4th Most Common Category  \
0      Dumpling Restaurant                    Diner           Discount Store   

  5th Most Common Category  
0                  Dog Run

#### Cluster 3

In [40]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Category  \
14       Scarborough               2                     Park   
20        North York               2                     Park   
23        North York               2                     Park   
25        North York               2        Food & Drink Shop   
44   Central Toronto               2                     Park   
50  Downtown Toronto               2                     Park   
74              York               2                     Park   
90         Etobicoke               2                    River   
98              York               2                     Park   

   2nd Most Common Category     3rd Most Common Category  \
14               Playground             Sculpture Garden   
20              Yoga Studio  Eastern European Restaurant   
23                     Bank            Convenience Store   
25                     Park                  Yoga Studio   
44              Swim School                     Bus Line   
50               Playground                        Trail   
74            Women's Store                       Market   
90                     Park                  Yoga Studio   
98        Convenience Store                  Yoga Studio   

       4th Most Common Category 5th Most Common Category  
14                  Yoga Studio                Drugstore  
20               Discount Store                  Dog Run  
23                          Bar              Yoga Studio  
25               Discount Store                  Dog Run  
44                  Yoga Studio                Drugstore  
50                     Building     Ethiopian Restaurant  
74         Fast Food Restaurant              Yoga Studio  
90                   Donut Shop       Dim Sum Restaurant  
98  Eastern European Restaurant           Discount Store

#### Cluster 4

In [41]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Category  \
63  Central Toronto               3                   Garden   

   2nd Most Common Category 3rd Most Common Category 4th Most Common Category  \
63                Drugstore       Dim Sum Restaurant                    Diner   

   5th Most Common Category  
63           Discount Store

#### Cluster 5

In [42]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Category  \
94  Etobicoke               4                     Bank   

   2nd Most Common Category 3rd Most Common Category 4th Most Common Category  \
94              Yoga Studio      Dumpling Restaurant           Discount Store   

   5th Most Common Category  
94                  Dog Run